### ISSUE TITLE
When OnEstablished callback takes a long time, peer may reset the connection with hold time expired

### ISSUE THREAD

**Issue by netixx** [jwhited/corebgp#55]
Hello,

To test a network device, I want to create 100 routers with the test device as peer and each send 100k routes to the device under test (DUT).

My code is roughly as followed:
- in a loop, I create 100 instances of the corebgp.Server
- I add the DUT peer to each server, passing a new instance of the plugin to each

I keep getting the following error messages after a little while:
```
2023/02/22 15:22:27 peer closed
write tcp 10.17.1.115:179->10.17.0.181:24379: write: broken pipe
```

On the router side, I can see:
```
RP/0/RP0/CPU0:2023 Feb 22 15:30:17.631 UTC: bgp[1069]: %ROUTING-BGP-5-ADJCHANGE_DETAIL : neighbor 10.17.1.115 Down - BGP Notification sent, hold time expired (VRF: default; AFI/SAFI: 1/1) (AS: xx) 
```

Looking further on the router, it actually doesn't receive any packet for 90 seconds, which is why it closes the connection.

Doing some debugging, it seems that the `writer.WriteUpdate` call blocks sometimes for a long time (90s or more).
More specifically the call to `u.conn.Write` in the following function:
```
func (u *updateMessageWriter) WriteUpdate(b []byte) error {
	/*
		https://tools.ietf.org/html/rfc4271#page-72
		Each time the local system sends a KEEPALIVE or UPDATE message, it
		restarts its KeepaliveTimer, unless the negotiated HoldTime value
		is zero.
	*/
	select {
	case <-u.closeCh:
		return io.ErrClosedPipe
	default:
		_, err := u.conn.Write(prependHeader(b, updateMessageType))
		if strings.HasPrefix(u.conn.LocalAddr().String(), "10.17.1.115") {
			fmt.Println("send done")
		}
		if err == nil {
			select {
			case <-u.closeCh:
			case u.resetKATimerCh <- struct{}{}:
			}
		}
		return err
	}
}
```

Note that I modified the function to add debug output, which leads to the following logs:
```
2023-02-22 17:41:34.928165013 +0000 UTC m=+277.043173280 send done
2023/02/22 17:41:34 send update done
2023/02/22 17:41:34 loop start route
2023/02/22 17:41:34 prepare route
2023/02/22 17:41:34 send update
2023-02-22 17:41:34.928208551 +0000 UTC m=+277.043216824 send done
2023/02/22 17:41:34 send update done
2023/02/22 17:41:34 loop start route
2023/02/22 17:41:34 prepare route
2023/02/22 17:41:34 send update
2023-02-22 17:41:34.928274148 +0000 UTC m=+277.043282415 send done
2023/02/22 17:41:34 send update done
2023/02/22 17:41:34 loop start route
2023/02/22 17:41:34 prepare route
2023/02/22 17:41:34 send update
```

There is no more logs until after the device closes the connection.

I don't get the issue when I use only one (or two) servers.
Could there be a concurrency issue that is limiting the number of simultaneous servers that can be run ?

Sorry for the vague report, I don't fully understand it myself... At this point I am not sure where the issue is exactly (could even be on the DUT...).

**Comment by jwhited** [comment]
> in a loop, I create 100 instances of the corebgp.Server
    I add the DUT peer to each server, passing a new instance of the plugin to each

Is there a particular reason you are creating 100 instances of `Server` instead of 100 peers on 1 `Server`?

> I add the DUT peer to each server, passing a new instance of the plugin to each

Can you share the plugin you're using?

> Doing some debugging, it seems that the writer.WriteUpdate call blocks sometimes for a long time (90s or more).
More specifically the call to u.conn.Write in the following function:

Can you `kill -ABRT <pid>` when it gets into this state? That should give you a goroutine dump to confirm it's blocked where you think it is.

**Comment by netixx** [comment]
> Is there a particular reason you are creating 100 instances of Server instead of 100 peers on 1 Server?

Yes, I need a single device under test to have 100 neighbors, so each server I create has the same (single) peer (same remote address, same remote router id)

> Can you share the plugin you're using?

yes, here is my main (sorry for the mess).
```
package main

import (
	"context"
	"encoding/hex"
	"flag"
	"fmt"
	"log"
	"math/rand"
	"net"
	"net/netip"
	"os"
	"os/signal"
	"runtime/pprof"
	"sync"
	"syscall"

	"github.com/jwhited/corebgp"
	"golang.org/x/sync/semaphore"
	"golang.org/x/sys/unix"
)

var (
	profileFile   = flag.String("cpuprofile", "", "write cpu profile to file")
	dfz           = LoadRouteDistribution("dfz.yml")
	sendSemaphore = semaphore.NewWeighted(5)
)

func randAsPath() []AsPathParamInterface {
	length := rand.Int31n(6)
	asPath := make([]AsPathParamInterface, length)
	var i int32
	for i = 0; i < length; i++ {
		var asn uint32
		// more 2 byte AS that 4 byte AS
		if rand.Intn(10) > 8 {
			for asn == 0 || asn == 23456 || asn == 65535 || (asn >= 65536 && asn <= 131071) {
				asn = uint32(rand.Int31n(401309))
			}
		} else {
			for asn == 0 || asn == 23456 || asn == 65535 {
				asn = uint32(rand.Int31n(65535))
			}
		}
		// if asn > math.MaxUint16 {
		asPath[i] = NewAs4PathParam(BGP_ASPATH_ATTR_TYPE_SEQ, []uint32{asn})
		// } else {
		// 	asPath[i] = NewAsPathParam(BGP_ASPATH_ATTR_TYPE_SEQ, []uint16{uint16(asn)})
		// }
	}
	return asPath
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func (p *plugin) sendDFZ(
	ctx context.Context,
	writer corebgp.UpdateMessageWriter,
	routes []*netip.Prefix,
	nextHops []netip.Addr,
	nlriBatch int,
	peer corebgp.PeerConfig,
) {
	// if err := sendSemaphore.Acquire(ctx, 1); err != nil {
	// 	log.Fatal(err)
	// 	return
	// }
	// defer sendSemaphore.Release(1)
	if p.id == 15 {
		log.Println("sending DFZ")
	}
	printPerc := 0.0
	for i := 0; i < len(routes); i += nlriBatch {
		if p.id == 15 {
			log.Println("loop start route")
		}
		nextHop := nextHops[rand.Intn(len(nextHops))]
		if perc := float64(i) / float64(len(routes)) * 100; perc > printPerc {
			// log.Printf("send %d/%d %.0f%%\n", i, len(routes), perc)
			printPerc += 10.0
		}
		select {
		case <-ctx.Done():
			return
		default:
			if p.id == 15 {
				log.Println("prepare route")
			}
			nhArr := nextHop.AsSlice()
			var commPrefix uint32 = 65512 << 16
			var commSuffix uint32 = uint32(nhArr[len(nhArr)-1])<<8 + uint32(nhArr[len(nhArr)-2])
			nlris := make([]*IPAddrPrefix, min(4, len(routes)-i))
			for batch := 0; batch < nlriBatch && i+batch < len(routes); batch++ {
				nlris[batch] = NewIPAddrPrefix(*routes[i+batch])
			}
			updateO := &BGPUpdate{
				WithdrawnRoutes: nil,
				PathAttributes: []PathAttributeInterface{
					NewPathAttributeOrigin(BGP_ORIGIN_ATTR_TYPE_IGP),
					// NewPathAttributeAsPath([]AsPathParamInterface{NewAs4PathParam(2, []uint32{65001})}),
					NewPathAttributeAsPath(randAsPath()),
					NewPathAttributeNextHop(nextHop),
					NewPathAttributeLocalPref(100),
					NewPathAttributeCommunities([]uint32{commPrefix + commSuffix}),
					NewPathAttributeMultiExitDisc(uint32(i % 100)),
				},
				NLRI: nlris,
			}

			updateVal, err := updateO.Serialize()
			if err != nil {
				fmt.Println(err)
				continue
			}
			if p.id == 15 {
				log.Printf("send update %d\n", len(updateVal))
			}
			err = writer.WriteUpdate(updateVal)
			if err != nil {
				log.Println(err)
				return
			}
			if p.id == 15 {
				log.Println("send update done")
			}
		}
	}
	if p.id == 15 {
		log.Println("send dfz done")
	}
}

type plugin struct {
	ctx    context.Context
	id     byte
	routes []*netip.Prefix
}

func (p *plugin) GetCapabilities(c corebgp.PeerConfig) []corebgp.Capability {
	caps := []corebgp.Capability{
		{
			Code: corebgp.CAP_ADD_PATH,
		},
		{
			Code: corebgp.CAP_FOUR_OCTET_AS,
		},
	}
	return caps
}

func (p *plugin) OnOpenMessage(peer corebgp.PeerConfig, routerID netip.Addr, capabilities []corebgp.Capability) *corebgp.Notification {
	return nil
}

func (p *plugin) sendRoutes(writer corebgp.UpdateMessageWriter, peer corebgp.PeerConfig) {
	select {
	case <-p.ctx.Done():
		return
	default:
		var nPB byte = 5
		nextHops := make([]netip.Addr, nPB)
		var i byte
		for i = 0; i < nPB; i++ {
			nextHops[i] = netip.AddrFrom4([4]byte{10, 11, p.id, i})
		}
		p.sendDFZ(
			p.ctx,
			writer,
			p.routes,
			nextHops,
			4,
			peer,
		)

		// send End-of-Rib
		log.Println("sending EOR")
		err := writer.WriteUpdate([]byte{
			0, // withdraw length
			//, // withdrawn NLRI
			0, // attribute length
			0, // attributes
			0, // NLRIs
		})
		if err != nil {
			fmt.Println(err)
		}
	}
}

func (p *plugin) OnEstablished(peer corebgp.PeerConfig, writer corebgp.UpdateMessageWriter) corebgp.UpdateMessageHandler {
	if p.id == 15 {
		log.Println("peer established")
	}
	p.sendRoutes(writer, peer)
	return p.handleUpdate
}

func (p *plugin) OnClose(peer corebgp.PeerConfig) {
	log.Println("peer closed")
}

func (p *plugin) handleUpdate(peer corebgp.PeerConfig, u []byte) *corebgp.Notification {
	log.Printf("got update message of len: %d for %d: %s", len(u), p.id, hex.Dump(u))
	return nil
}

func main() {
	flag.Parse()
	if *profileFile != "" {
		f, err := os.Create(*profileFile)
		if err != nil {
			log.Fatal(err)
		}
		if pprof.StartCPUProfile(f) != nil {
			panic(err)
		}
		defer pprof.StopCPUProfile()
	}

	scaleFactor := 1.5
	var numberOfPeers byte = 50
	currentRRIPv4RouteCount := 7438625
	currentRRIPv4BestPathCount := 692304

	dfzRoutesCount := dfz.IPv4.Total()

	targetBestPathRouteCount := int(scaleFactor * float64(currentRRIPv4BestPathCount))
	dfzScaleFactor := float64(targetBestPathRouteCount) / float64(dfzRoutesCount)

	fmt.Printf("DFZ scale factor is %.2f", dfzScaleFactor)

	// of those routes, we need currentRRIPv4BestPathCount * scaleFactor to be bestpaths
	// We know that all DFZ route will be a best-path, to simplify, send a total of currentRRIPv4BestPathCount * scaleFactor routes
	// but have overlap between the peers to reach the desired currentRRIPv4RouteCount * scaleFactor

	srvErrCh := make(chan error)
	ctxC, cancel := context.WithCancel(context.Background())

	// iface, _ := netlink.LinkByName("ens7")
	// peerIP := netip.MustParseAddr("10.17.0.181")
	peerIP := netip.MustParseAddr("10.17.0.184")
	var i byte
	servers := make([]*corebgp.Server, numberOfPeers)

	// each device needs to send
	// currentRRIPv4RouteCount * scaleFactor/numberOfPeers
	// taken amongst currentRRIPv4BestPathCount * scaleFactor routes
	routes := dfz.GenerateV4(dfzScaleFactor)
	routesPerPeer := int(float64(currentRRIPv4RouteCount) * scaleFactor / float64(numberOfPeers))
	wg := &sync.WaitGroup{}
	log.Printf("Sending %d routes per peer\n", routesPerPeer)
	for i = 0; i < numberOfPeers; i++ {
		routerID := netip.AddrFrom4([4]byte{10, 17, 1, byte(100 + i)})
		// netAddr, _ := netlink.ParseAddr(fmt.Sprintf("%s/%d", routerID.String(), 24))
		// netlink.AddrAdd(iface, netAddr)
		log.Printf("starting server %d\n", i)
		servers[i] = serveAndSend(
			ctxC,
			wg,
			srvErrCh,
			i,
			routerID,
			peerIP,
			randomRoutes(
				routes,
				routesPerPeer,
			),
		)
	}

	log.Println("done starting servers")

	sigCh := make(chan os.Signal, 1)
	signal.Notify(sigCh, syscall.SIGINT, syscall.SIGTERM)
	select {
	case <-sigCh:
		log.Println("got signal")
		break
		// srv.Close()
		// <-srvErrCh
	case err := <-srvErrCh:
		log.Fatalf("serve error: %v", err)
		break
	}
	cancel()
	wg.Wait()
}

func control(network, address string, c syscall.RawConn) error {
	c.Control(func(fd uintptr) {
		if err := unix.SetsockoptInt(int(fd), unix.SOL_SOCKET, unix.SO_RCVBUF, 1024); err != nil {
			fmt.Printf("Set socket receive buffer size failed: %v\n", err)
		}
		if err := unix.SetsockoptInt(int(fd), unix.SOL_SOCKET, unix.SO_SNDBUF, 1024); err != nil {
			fmt.Printf("Set socket send buffer size failed: %v\n", err)
		}

		// if err := syscall.SetsockoptInt(int(fd), syscall.IPPROTO_TCP, syscall.TCP_KEEPINTVL, 10); err != nil {
		// 	fmt.Printf("Set socket keepalive interval failed: %v\n", err)
		// }
		// if err := syscall.SetsockoptInt(int(fd), syscall.IPPROTO_TCP, syscall.TCP_KEEPIDLE, 10); err != nil {
		// 	fmt.Printf("Set socket keepalive idle failed: %v\n", err)
		// }
	})
	return nil
}

func randomRoutes(routes []*netip.Prefix, length int) []*netip.Prefix {
	ret := make([]*netip.Prefix, length)

	for i := range ret {
		ret[i] = routes[rand.Intn(len(routes))]
	}

	return ret
}

func serveAndSend(ctx context.Context, wg *sync.WaitGroup, srvErrCh chan error, id byte, localIP, remoteIP netip.Addr, routes []*netip.Prefix) *corebgp.Server {
	srv, err := corebgp.NewServer(localIP)
	if err != nil {
		log.Fatalf("error constructing server: %v", err)
	}

	p := &plugin{
		ctx:    ctx,
		id:     id,
		routes: routes,
	}
	err = srv.AddPeer(
		corebgp.PeerConfig{
			RemoteAddress: remoteIP,
			LocalAS:       xxx,
			RemoteAS:      xxx,
		},
		p,
		corebgp.WithLocalAddress(localIP),
		corebgp.WithPassive(),
	)

	if err != nil {
		log.Fatalf("error adding peer: %v", err)
	}

	lc := &net.ListenConfig{
		Control: control,
	}

	lis, err := lc.Listen(ctx, "tcp", fmt.Sprintf("%s:179", localIP.String()))
	if err != nil {
		log.Fatalf("error constructing listener: %v", err)
	}

	// server thread
	go func() {
		err := srv.Serve([]net.Listener{lis})
		srvErrCh <- err
	}()
	wg.Add(1)
	// killer thread
	go func() {
		<-ctx.Done()
		log.Printf("Killing server %d\n", id)
		srv.Close()
		log.Printf("Server %d finished\n", id)
		wg.Done()
	}()

	return srv
}
```

> Can you kill -ABRT <pid> when it gets into this state? That should give you a goroutine dump to confirm it's blocked where you think it is.

Here is I think a example of a goroutine that is blocked:
```
goroutine 479 [IO wait, 2 minutes]:
runtime.gopark(0x7f9a3710a408?, 0xb?, 0x1?, 0x0?, 0x62?)
        /home/ubuntu/sdk/go1.20.1/src/runtime/proc.go:381 +0xd6 fp=0xc0088856d0 sp=0xc0088856b0 pc=0x436b56
runtime.netpollblock(0x493245?, 0x403faf?, 0x0?)
        /home/ubuntu/sdk/go1.20.1/src/runtime/netpoll.go:527 +0xf7 fp=0xc008885708 sp=0xc0088856d0 pc=0x42f5b7
internal/poll.runtime_pollWait(0x7f9a377679b8, 0x77)
        /home/ubuntu/sdk/go1.20.1/src/runtime/netpoll.go:306 +0x89 fp=0xc008885728 sp=0xc008885708 pc=0x460ee9
internal/poll.(*pollDesc).wait(0xc009c56300?, 0xc009d17b74?, 0x0)
        /home/ubuntu/sdk/go1.20.1/src/internal/poll/fd_poll_runtime.go:84 +0x32 fp=0xc008885750 sp=0xc008885728 pc=0x4ab032
internal/poll.(*pollDesc).waitWrite(...)
        /home/ubuntu/sdk/go1.20.1/src/internal/poll/fd_poll_runtime.go:93
internal/poll.(*FD).Write(0xc009c56300, {0xc009d17b60, 0x5c, 0x60})
        /home/ubuntu/sdk/go1.20.1/src/internal/poll/fd_unix.go:391 +0x2f6 fp=0xc008885800 sp=0xc008885750 pc=0x4ac276
net.(*netFD).Write(0xc009c56300, {0xc009d17b60?, 0x44c036?, 0x60?})
        /home/ubuntu/sdk/go1.20.1/src/net/fd_posix.go:96 +0x29 fp=0xc008885848 sp=0xc008885800 pc=0x5126a9
net.(*conn).Write(0xc009c58048, {0xc009d17b60?, 0x562920?, 0xc009d17b60?})
        /home/ubuntu/sdk/go1.20.1/src/net/net.go:195 +0x45 fp=0xc008885890 sp=0xc008885848 pc=0x51c8e5
net.(*TCPConn).Write(0xc009adfec0?, {0xc009d17b60?, 0x2?, 0x2?})
        <autogenerated>:1 +0x29 fp=0xc0088858c0 sp=0xc008885890 pc=0x526ee9
github.com/jwhited/corebgp.(*updateMessageWriter).WriteUpdate(0xc009515720, {0xc009d1e980, 0x49, 0x0?})
        /home/ubuntu/gohack/github.com/jwhited/corebgp/fsm.go:798 +0x12c fp=0xc008885980 sp=0xc0088858c0 pc=0x54246c
main.(*plugin).sendDFZ(0xc000b39680, {0x5c7ab8, 0xc00006e1e0}, {0x5c7180, 0xc009515720}, {0xc00600c000, 0x367b6, 0xa?}, {0xc00097ac00, 0x5, ...}, ...)
        /home/ubuntu/corebgp/main.go:123 +0x827 fp=0xc008885b58 sp=0xc008885980 pc=0x54c767
main.(*plugin).sendRoutes(0xc000b39680, {0x5c7180, 0xc009515720}, {{{0x7f9a5eb4d108?, 0x20?}, 0xc000010030?}, 0x9515720?, 0xc0?})
        /home/ubuntu/corebgp/main.go:171 +0x1b0 fp=0xc008885c28 sp=0xc008885b58 pc=0x54cc70
main.(*plugin).OnEstablished(0xc000b39680, {{{0x0?, 0x560fe0?}, 0xc000010030?}, 0x4633ed?, 0x0?}, {0x5c7180?, 0xc009515720?})
        /home/ubuntu/corebgp/main.go:199 +0x105 fp=0xc008885ca0 sp=0xc008885c28 pc=0x54cec5
github.com/jwhited/corebgp.(*fsm).established.func2()
        /home/ubuntu/gohack/github.com/jwhited/corebgp/fsm.go:846 +0x172 fp=0xc008885e30 sp=0xc008885ca0 pc=0x542972
github.com/jwhited/corebgp.(*fsm).established(0xc00d1526e0)
        /home/ubuntu/gohack/github.com/jwhited/corebgp/fsm.go:955 +0x132 fp=0xc008885eb8 sp=0xc008885e30 pc=0x542712
github.com/jwhited/corebgp.(*fsm).run(0xc00d1526e0)
        /home/ubuntu/gohack/github.com/jwhited/corebgp/fsm.go:168 +0x288 fp=0xc008885fc8 sp=0xc008885eb8 pc=0x53f408
github.com/jwhited/corebgp.(*fsm).start.func1()
        /home/ubuntu/gohack/github.com/jwhited/corebgp/fsm.go:186 +0x26 fp=0xc008885fe0 sp=0xc008885fc8 pc=0x53f6e6
runtime.goexit()
        /home/ubuntu/sdk/go1.20.1/src/runtime/asm_amd64.s:1598 +0x1 fp=0xc008885fe8 sp=0xc008885fe0 pc=0x466361
created by github.com/jwhited/corebgp.(*fsm).start
        /home/ubuntu/gohack/github.com/jwhited/corebgp/fsm.go:186 +0x56
```


I did some further testing:
- With another device, I don't get exactly the same packet behaviour in the packet capture, but I still get some hold timer expired.
- I did a test with much less routes per peer (around 10k), and it also leads to some hold timer expiring.
- I played around with SO_SNDBUF values. If I don't set it, I do get 1 TCP packet per update (because of TCP_NODELAY being true by default), however I still get hold timer expired.
- What is weird is that I get this only for a few peers in the 50 peers (I reduced it to make testing easier).
- Also it seems that in some cases, the session cannot go up again.

Lastly :), looking at this bit of code
```
handler := f.peer.plugin.OnEstablished(f.peer.config, writer)

		for {
			select {
			case <-f.closeCh:
				n := newNotification(NOTIF_CODE_CEASE, 0, nil)
				f.sendNotification(n) // nolint: errcheck
				return disabledState, newNotificationError(n, true)
			case <-f.holdTimer.C:
				n := newNotification(NOTIF_CODE_HOLD_TIMER_EXPIRED, 0, nil)
				f.sendNotification(n) // nolint: errcheck
				return idleState, newNotificationError(n, true)
			case <-f.keepAliveTimer.C:
				err := f.sendKeepAlive()
```

it seem we don't handle the received message (and send the keepalive) untile the OnEstablished callback has returned (note that I tried to make a goroutine inside so that it returns immediately, but this made things worst - a lot more hold time expired).

**Comment by jwhited** [comment]
> I played around with SO_SNDBUF values. If I don't set it, I do get 1 TCP packet per update (because of TCP_NODELAY being true by default), however I still get hold timer expired.

The goroutine does look like it's blocked on writing to the socket. Why are you attempting to set such a small snd/rcv buff?

> Also it seems that in some cases, the session cannot go up again.

Take a look at [`corebgp.SetLogger()`](https://pkg.go.dev/github.com/jwhited/corebgp#SetLogger). This will give you some more information on the fsm state and peer damping.

> it seem we don't handle the received message (and send the keepalive) untile the OnEstablished callback has returned (note that I tried to make a goroutine inside so that it returns immediately, but this made things worst - a lot more hold time expired).

Is your `OnEstablished()` particularly heavy? It would have to block for a number of seconds for this order to matter.

If you tcpdump do you see missing keepalives?

**Comment by jwhited** [comment]
```
func (p *plugin) OnEstablished(peer corebgp.PeerConfig, writer corebgp.UpdateMessageWriter) corebgp.UpdateMessageHandler {
	if p.id == 15 {
		log.Println("peer established")
	}
	p.sendRoutes(writer, peer)
	return p.handleUpdate
}
```

So you're trying to send routes prior to returning in `OnEstablished()`. The small snd/rcv buf is probably making this problem worse. I'll have to think more about this ordering, but you should be able to modify the i/o model to make this run in a separate goroutine.

**Comment by jwhited** [comment]
You don't want to block for a long period of time in `OnEstablished()`. It needs to return an inbound update message handler for the state machine to use while in an established state. 

Here's a simplified `Plugin` example that creates a separate goroutine to send outbound update messages when `OnEstablished()` fires. You can count on `OnClose()` firing when the peer transitions out of an established state, so the goroutine will not leak:

```go
type plugin struct {
	ctx      context.Context
	cancelFn context.CancelFunc
	wg       *sync.WaitGroup
}

func newPlugin() *plugin {
	return &plugin{
		wg: &sync.WaitGroup{},
	}
}

func (p *plugin) GetCapabilities(c corebgp.PeerConfig) []corebgp.Capability {
	return nil
}

func (p *plugin) OnOpenMessage(peer corebgp.PeerConfig, routerID netip.Addr, capabilities []corebgp.Capability) *corebgp.Notification {
	log.Println("open message received")
	return nil
}

func (p *plugin) sendRoutes(writer corebgp.UpdateMessageWriter) {
	defer p.wg.Done()
	for {
		select {
		case <-p.ctx.Done():
			return
		default:
			update := make([]byte, 0)
			// construct update message
			err := writer.WriteUpdate(update)
			if err != nil {
				return
			}
		}
	}
}

func (p *plugin) OnEstablished(peer corebgp.PeerConfig, writer corebgp.UpdateMessageWriter) corebgp.UpdateMessageHandler {
	log.Println("peer established")
	p.ctx, p.cancelFn = context.WithCancel(context.Background())
	p.wg.Add(1)
	go p.sendRoutes(writer)
	return p.handleUpdate
}

func (p *plugin) OnClose(peer corebgp.PeerConfig) {
	p.cancelFn()
	p.wg.Wait()
	log.Println("peer closed")
}

func (p *plugin) handleUpdate(peer corebgp.PeerConfig, b []byte) *corebgp.Notification {
	log.Printf("got update message: %x", b)
	return nil
}
```

There are other ways to achieve something similar, this is just one example.

**Comment by netixx** [comment]
I had tried the goroutine in the OnEstablished before and it didn't work then.

I tried again today with your code, and I couldn't reproduce the issue anymore.
Looks like the DUT router was causing issues. In my capture I could see TCP zero window from the router, which blocked TCP exchange altogether...

If I may, the example in the docs could be updated with the example posted here. Or the code could be changed to make sure that the OnEstablished does not block further processing of the BGP message
